### PR TITLE
Fixed data race.

### PR DESCRIPTION
**Pull request by Frozen**
https://github.com/harmony-one/harmony/issues/4558
https://github.com/harmony-one/harmony/issues/4557
https://github.com/harmony-one/harmony/issues/4556
https://github.com/harmony-one/harmony/issues/4555

### CODE REVIEW

File: consensus/consensus_v2.go [hunk 1]
```diff
@@ -575,19 +575,19 @@ func (consensus *Consensus) preCommitAndPropose(blk *types.Block) error {
...
 				return
 			}
 		}
-
+		consensus.mutex.Lock()
 		consensus.getLogger().Info().Msg("[preCommitAndPropose] Start consensus timer")
 		consensus.consensusTimeout[timeoutConsensus].Start()
 
 		// Send signal to Node to propose the new block for consensus
 		consensus.getLogger().Info().Msg("[preCommitAndPropose] sending block proposal signal")
```
Comment (GheisMohammadi) [inline consensus/consensus_v2.go:589]: GetLogger() ?

File: consensus/consensus_v2.go [hunk 2]
```diff
@@ -575,19 +575,19 @@ func (consensus *Consensus) preCommitAndPropose(blk *types.Block) error {
...
-
+		consensus.mutex.Unlock()
```
Comment (GheisMohammadi) [inline consensus/consensus_v2.go:590]: defer?

File: consensus/consensus_v2.go (hunk 1 above)
Comment (Frozen) [reply consensus/consensus_v2.go:589]: No, we obtained lock above.

Comment (Frozen) [reply consensus/consensus_v2.go:589]: GetLogger() has lock inside, so it will be deadlock.

Comment (GheisMohammadi) [reply consensus/consensus_v2.go:589]: GetLogger is a public function, it should be improved to not have such a constraints. We can't make a function public and ask users to not use it in a certain conditions, right?

File: consensus/consensus_v2.go (hunk 2 above)
Comment (GheisMohammadi) [reply consensus/consensus_v2.go:590]: to excludes line  `consensus.ReadySignal(AsyncProposal)` ? any reason for that?

**Pull request merged**